Copied over the harmonization files

steveraysteveray opened the pull request.

steveraysteveray commented:
@fkleedorfer, would you be willing to run your conversion multiplier generator code on this branch? It contains a significant number of new units that lack conversion multipliers. Feel free to add an additional commit to this branch directly.

Thanks in advance!

fkleedorfer commented on vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:
```diff
@@ -7505,6 +8603,15 @@ unit:FT_H2O
   rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
   rdfs:label "Foot of Water"@en ;
 .
+<http://qudt.org/vocab/unit/FT_H2O_39.2DEG_F>
```
The `'.'` here trips up my generation of Java/Javascript constants (no dots allowed). Not a huge deal, but I noticed QUDT already has a rule for commas in dimension vectors: replace `.` with `'pt'`. It would be consistent to apply this here, too.

fkleedorfer replied:
So, the suggestion would be to use `FT_H2O_39pt2DEG_F`

fkleedorfer replied:
@steveraysteveray is it ok if I make that change in my upcoming commit(s) to this branch?

fkleedorfer commented on vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:
```diff
@@ -7505,6 +8603,15 @@ unit:FT_H2O
+  dcterms:description "non SI-conform unit of pressure according to the Anglo-American and Imperial system for units, whereas the value of 1 ftH₂O is equivalent to the static pressure, which is generated by a head of water at a temperature 39.2°F with a height of 1 foot" ;
+  qudt:hasDimensionVector qkdv:A0E0L-1I0M1H0T-2D0 ;
+  qudt:iec61360Code "0112/2///62720#UAB237" ;
+  qudt:symbol "ftH₂O (39,2 ºF)" ;
```
The `,` as the decimal separator is surprising here

fkleedorfer commented:
I did the conversionMultiplier generation, please cross-check. 

Of special note:
* the commit contains some corrections of (presumably) wrong conversion multipliers. However, the original multipliers have not been removed. Someone needs to cross check the corrections and remove whichever one is actually wrong.

* many corrections concern units derived from `'BIT'` or `'BYTE'` -  I am especially unsure about those. Please double check!

* If you can provide more information for GAL2_US, we can come up with the multiplier of derived units.

steveraysteveray commented:
@fkleedorfer, thanks so much for your rapid response! Let me address some of the issues you observed:

1. . vs 'pt' vs 'dot'.   The QUDT Board recently revisited this topic, because all of the encodings currently exist at times. We agreed to go with 'dot' instead of '.' in all the URIs where it shows up. (So, FT_H2O_39dot2DEG_F)
2. We have still not discussed and agreed on your suggestion in #815, so please stay tuned on that.
3. We are trying to stick with the decimal . in places where you have found a ,  so yes, please feel free to fix those. (qudt:symbol "ftH₂O (39.2 ºF)" ;
4. Regarding the units derived from BIT and BYTE, we defined the conversion multiplier of unit:BIT in terms of unit:NAT, which is the unit of information entropy based on natural logarithms, consistent with thermodynamic entropy. So we should double-check all the derived units building on BIT and BYTE for consistency with this.
5. Regarding GAL2_US, that means square US gallons. Not sure what the ambiguity is here.

If you would like, we could divide up the tasks. Let me know which ones do not easily automate for you and I will pick those up. We can continue in this branch.

Thanks again.

fkleedorfer commented:
> @fkleedorfer, thanks so much for your rapid response! Let me address some of the issues you observed:
> 
>     1. . vs 'pt' vs 'dot'.   The QUDT Board recently revisited this topic, because all of the encodings currently exist at times. We agreed to go with 'dot' instead of '.' in all the URIs where it shows up. (So, FT_H2O_39dot2DEG_F)

Just ran a search, there will be lots of changes. I'll just touch the one that bit me.

> 
>     2. We have still not discussed and agreed on your suggestion in [Change request to avoid ambiguous use of hyphen introduced in 2.1.30 #815](https://github.com/qudt/qudt-public-repo/issues/815), so please stay tuned on that.
> 

Understood, no worries. Sorry again for making a fuss after the fact.

>     3. We are trying to stick with the decimal . in places where you have found a ,  so yes, please feel free to fix those. (qudt:symbol "ftH₂O (39.2 ºF)" ;
> 

ok, will do.

>     4. Regarding the units derived from BIT and BYTE, we defined the conversion multiplier of unit:BIT in terms of unit:NAT, which is the unit of information entropy based on natural logarithms, consistent with thermodynamic entropy. So we should double-check all the derived units building on BIT and BYTE for consistency with this.
> 

ok, then I'll add those corrections again and we can revisit. 

>     5. Regarding GAL2_US, that means square US gallons. Not sure what the ambiguity is here.
> 

oh, then it's a mistake, I believe - it should be unit:GAL_US2 and unit:GAL_UK2. Unless you protest, I'll change it

> 
> If you would like, we could divide up the tasks. Let me know which ones do not easily automate for you and I will pick those up. We can continue in this branch.

Let me just finish the points I noted above (unless you have objections) in this branch (1hr from now I estimate).

steveraysteveray commented:
1. I don't see so many occurrences in the URIs, but we can dig deeper after your pass.
2. Agreed
3. Agreed
4. Agreed
5. Agreed

steveraysteveray commented:
I have checked all the erroneous conversion multipliers, and you are of course correct!

fkleedorfer commented:
I've done all I promised. Have you checked the multipliers from the last commit and the dimension vectors as well?

steveraysteveray commented:
I'm working on checking the incorrect dimension vectors and added some of relevant quantity kinds when it seemed to make sense. I did not change any of the incorrect dimension vectors, so this still needs to be done.

Progress so far:

The GigaFLOPS-PER-SEC and related units are just wrong and should not exist with -PER-SEC. I have removed those units. (Similar to the common phrase "ATM Machine" where ATM already means "Automated Teller Machine"!)

(Incidentally, in your code, the supporting comment line seemed wrong in this case:   # factor unit tree for unit g/(cm³.bar).
The cm³ should have been just cm, but the calculations are still correct.)

steveraysteveray commented:
@fkleedorfer, I believe I have now deleted the incorrect, and added the corrected, dimension vectors, along with adding the corrected conversion multipliers (since you had already deleted the incorrect ones). It would be good if you could run your routines once more to ensure nothing slipped through.

Other than that, when I run my validator, I get only 2 kinds of errors: 

1. Some quantity kinds that are missing dimension vectors (which will be a manual process)
2. Some units that are missing quantity kinds (also a manual process)

fkleedorfer commented:
Ran the conversion multiplier checks again, came up with some more corrections/additions (after some refactoring of the code). sorry for that.

fkleedorfer commented:
Also I found `unit:PER-MILLE-PER-PSI`, which does not conform to the convention. I deprecated it in favor of new `unit:PER-MILLE-PSI`. Hope that's correct and ok

steveraysteveray commented on vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:
```diff
@@ -13902,7 +13906,11 @@ unit:KiloBYTE
-  qudt:conversionMultiplier 5545.17744447956 ;
+  # WRONG MULTIPLIER  : unit:KiloBYTE - calculated from factors: 5678.261703147071974745965538985370, actual: 5545.17744447956
+  #
+  #  ──┬ unit:KiloBYTE multiplier: 5545.17744447956 (correct: 5678.26170314707197474596553898537)
```
If BYTE is 5.545... why isn't KiloBYTE 5545...?

steveraysteveray replied:
Are you conflating KiloBYTE with KibiBYTE?

fkleedorfer replied:
Thanks for spotting this. No clue how that happened.

fkleedorfer replied:
It was caused by `unit:KiloBYTE` having the `prefix:Kibi` instead of `prefix:Kilo`. Fixed in subsequent commit.

steveraysteveray replied:
Do you depend on qudt:prefix? I notice that it is not consistently present.

steveraysteveray replied:
Also, the use of qudt:prefix would be ambiguous with compound units such as unit:KiloGM-PER-MegaBTU_IT.

fkleedorfer commented:
I added symbols and ucum codes. I know it's a lot, so if you don't want them, just remove the commits. 

I think I am done. Let me know if you need anything else.

steveraysteveray commented on vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:
```diff
@@ -39977,3 +39985,9459 @@ unit:PERCENT-PER-MO qudt:symbol "%/mo" .
+##         UCUM CODES
+##
+######################################################################################################
+  # MISSING UCUMCODE: unit:AT - calculated from factors: A
```
These two UCUM codes involving Ampere are not correct.

steveraysteveray commented on vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:
```diff
@@ -39977,3 +39985,9459 @@ unit:PERCENT-PER-MO qudt:symbol "%/mo" .
+  #          └─── unit:GM ucumCode: g
+unit:LB-PER-LB qudt:ucumCode "[lb_av].[lb_av]-1"^^<http://qudt.org/schema/qudt/UCUMcs> .
+
+  # MISSING UCUMCODE: unit:LunarMass - calculated from factors: kg
```
Can't be right.

fkleedorfer replied:
I weeded out a number of such cases, but apparently, there are still some in there. Let me go over the algorithm again.

fkleedorfer replied:
I'll adapt it such that no ucum code/symbol is generated for a unit without hyphen in the localname.

steveraysteveray replied:
OK, sounds better!

fkleedorfer replied:
done

steveraysteveray reviewed (commented):
It seems that your algorithm ignores any factor units it does not recognize, resulting in a faulty UCUM code. Perhaps an improvement would be to simply not recommend a code in those cases.

fkleedorfer commented:
Re-did the symbols/ucum codes. Looks better now.

jhodgesatmb commented:
I do not see anything to review anymore for #855.

On Mon, Feb 26, 2024 at 8:39 AM QUDT ***@***.***> wrote:

> Merged #855 <https://github.com/qudt/qudt-public-repo/pull/855> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qudt/qudt-public-repo/pull/855#event-11924218803>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATQRWPHUSQDCEXCWV4JA53YVS3DDAVCNFSM6AAAAABCRYYQSKVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJRHEZDIMRRHA4DAMY>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 
Jack

steveraysteveray commented:
Correct. The publication script is running as we speak.
